source/extensions/transport_sockets/alts/tsi_frame_protector.cc

```diff
@@ -1,73 +1,119 @@
+#include "src/core/tsi/transport_security_interface.h"
+
 namespace Envoy {
 namespace Extensions {
 namespace TransportSockets {
 namespace Alts {
 
-// TODO(lizan): tune size later
-static constexpr uint32_t BUFFER_SIZE = 16384;
-
 TsiFrameProtector::TsiFrameProtector(CFrameProtectorPtr&& frame_protector)
     : frame_protector_(std::move(frame_protector)) {}
 
 tsi_result TsiFrameProtector::protect(Buffer::Instance& input, Buffer::Instance& output) {
   ASSERT(frame_protector_);
 
-  unsigned char protected_buffer[BUFFER_SIZE];
-  while (input.length() > 0) {
-    auto* message_bytes = reinterpret_cast<unsigned char*>(input.linearize(input.length()));
-    size_t protected_buffer_size = BUFFER_SIZE;
-    size_t processed_message_size = input.length();
-    tsi_result result =
-        tsi_frame_protector_protect(frame_protector_.get(), message_bytes, &processed_message_size,
-                                    protected_buffer, &protected_buffer_size);
-    if (result != TSI_OK) {
-      ASSERT(result != TSI_INVALID_ARGUMENT && result != TSI_UNIMPLEMENTED);
-      return result;
-    }
-    output.add(protected_buffer, protected_buffer_size);
-    input.drain(processed_message_size);
+  if (input.length() == 0) {
+    return TSI_OK;
+  }
+
+  grpc_core::ExecCtx exec_ctx;
+  grpc_slice input_slice = grpc_slice_from_copied_buffer(
+      reinterpret_cast<char*>(input.linearize(input.length())), input.length());
+
+  grpc_slice_buffer message_buffer;
+  grpc_slice_buffer_init(&message_buffer);
+  grpc_slice_buffer_add(&message_buffer, input_slice);
+
+  grpc_slice_buffer protected_buffer;
+  grpc_slice_buffer_init(&protected_buffer);
+
+  tsi_result result = tsi_zero_copy_grpc_protector_protect(frame_protector_.get(), &message_buffer,
+                                                           &protected_buffer);
+
+  if (result != TSI_OK) {
+    ASSERT(result != TSI_INVALID_ARGUMENT && result != TSI_UNIMPLEMENTED);
+    grpc_slice_buffer_destroy(&message_buffer);
+    grpc_slice_buffer_destroy(&protected_buffer);
+    return result;
   }
 
-  // TSI may buffer some of the input internally. Flush its buffer to protected_buffer.
-  size_t still_pending_size;
-  do {
-    size_t protected_buffer_size = BUFFER_SIZE;
-    tsi_result result = tsi_frame_protector_protect_flush(
-        frame_protector_.get(), protected_buffer, &protected_buffer_size, &still_pending_size);
-    if (result != TSI_OK) {
-      ASSERT(result != TSI_INVALID_ARGUMENT && result != TSI_UNIMPLEMENTED);
-      return result;
-    }
-    output.add(protected_buffer, protected_buffer_size);
-  } while (still_pending_size > 0);
+  const size_t protected_data_length = protected_buffer.length;
+  char* protected_data = new char[protected_data_length];
+
+  grpc_slice_buffer_move_first_into_buffer(&protected_buffer, protected_data_length,
+                                           protected_data);
+
+  auto fragment = new Buffer::BufferFragmentImpl(
```

Comment:
Ah thanks for @jmarantz reminder, we might want to refactor `GrpcSliceBufferFragmentImpl` https://github.com/envoyproxy/envoy/blob/master/source/common/grpc/google_grpc_utils.cc#L92 and make it usable here?